### DISCUSSION TITLE
Accessing View Factor Study inside GrayDiffuseRadiation

### DISCUSSION THREAD

**Discussion by mrplants** [idaholab/moose#23696] [Q&A General] [answered]
I am using the GrayDiffuseRadiation action and I want to visualize the rays from the view factor study that I assume it uses internally.  How do I access this view factor study?

Here is the relevant portion of my HIT file.

```
[GrayDiffuseRadiation]
  [vacuum_radiation]
    boundary = 'stator_inner_surface_sideset rotor_outer_surface_sideset'
    emissivity = '0.03 0.03'
    n_patches = '20 20'
    temperature = Temperature
    ray_tracing_face_type = 'GAUSS'
  []
[]

...

[Outputs]
  exodus = true
  [rays]
    type = RayTracingExodus
    study = NOT_SURE_WHAT_TO_PUT_HERE
    execute_on = FINAL
  []
[]
```

**Comment by mrplants** [comment, marked as answer]
I am using the same code from this question here:  https://github.com/idaholab/moose/discussions/23590

**Reply by GiudGiud** [reply]
I suspect it should be this: `vacuum_radiation`

**Reply by mrplants** [reply]
Unfortunately, that does not work.  See attached screenshot for the results:
<img width="906" alt="Screenshot 2023-03-12 at 22 36 12" src="https://user-images.githubusercontent.com/3487464/224595955-2f0c969e-af5b-4b81-9bb8-c9d94ea69f3c.png">

**Reply by GiudGiud** [reply]
in the code I see ray_study_uo_ + _name

so try ray_study_uo_vacuum_radiation

If that does not work, add Debug/show_actions=true and look for the name of the study there

**Reply by mrplants** [reply]
Good idea!  With that name, the error is:

<img width="586" alt="Screenshot 2023-03-12 at 23 05 50" src="https://user-images.githubusercontent.com/3487464/224598931-337c0003-2c3e-45b8-ac01-9f99877f25c9.png">

**Reply by mrplants** [reply]
Maybe the study doesn't cache the rays by default?  Is there a way to force that through the GrayDiffuseRadiation action?

**Reply by mrplants** [reply]
Here's the debug output:

```
[DBG][ACT] TASK (           common_output) TYPE (              CommonOutputAction) NAME (         Outputs) Memory usage 57MB
[DBG][ACT] TASK ( setup_recover_file_base) TYPE (      SetupRecoverFileBaseAction) NAME (                ) Memory usage 57MB
[DBG][ACT] TASK (   check_copy_nodal_vars) TYPE (             CopyNodalVarsAction) NAME (     Temperature) Memory usage 57MB
[DBG][ACT] TASK (   check_copy_nodal_vars) TYPE (             CopyNodalVarsAction) NAME (                ) Memory usage 57MB
[DBG][ACT] TASK (              setup_mesh) TYPE (                 SetupMeshAction) NAME (            Mesh) Memory usage 57MB
[DBG][ACT] TASK (        add_geometric_rm) TYPE (    CreateDisplacedProblemAction) NAME (            Mesh) Memory usage 57MB
[DBG][ACT] TASK (        add_geometric_rm) TYPE (          AddRelationshipManager) NAME (                ) Memory usage 57MB
[DBG][ACT] TASK (      add_mesh_generator) TYPE (          AddMeshGeneratorAction) NAME (       file_mesh) Memory usage 57MB
[DBG][ACT] TASK (      add_mesh_generator) TYPE (          AddMeshGeneratorAction) NAME (rotor_outer_surface_sideset) Memory usage 57MB
[DBG][ACT] TASK (      add_mesh_generator) TYPE (          AddMeshGeneratorAction) NAME (stator_inner_surface_sideset) Memory usage 57MB
[DBG][ACT] TASK (   append_mesh_generator) TYPE (         RadiationTransferAction) NAME (vacuum_radiation) Memory usage 57MB
[DBG][ACT] TASK ( execute_mesh_generators) TYPE (           ExecuteMeshGenerators) NAME (                ) Memory usage 58MB
[DBG][ACT] TASK (       recover_meta_data) TYPE (      SetupRecoverFileBaseAction) NAME (                ) Memory usage 67MB
[DBG][ACT] TASK (           set_mesh_base) TYPE (                 SetupMeshAction) NAME (            Mesh) Memory usage 67MB
[DBG][ACT] TASK (     attach_geometric_rm) TYPE (          AddRelationshipManager) NAME (                ) Memory usage 67MB
[DBG][ACT] TASK (               init_mesh) TYPE (                 SetupMeshAction) NAME (            Mesh) Memory usage 67MB
[DBG][ACT] TASK (            prepare_mesh) TYPE (         SetupMeshCompleteAction) NAME (            Mesh) Memory usage 67MB
[DBG][ACT] TASK (     uniform_refine_mesh) TYPE (         SetupMeshCompleteAction) NAME (            Mesh) Memory usage 68MB
[DBG][ACT] TASK (     setup_mesh_complete) TYPE (         SetupMeshCompleteAction) NAME (            Mesh) Memory usage 68MB
[DBG][ACT] TASK (     setup_mesh_complete) TYPE (         RadiationTransferAction) NAME (vacuum_radiation) Memory usage 68MB
[DBG][ACT] TASK (   determine_system_type) TYPE (      CreateProblemDefaultAction) NAME (                ) Memory usage 68MB
[DBG][ACT] TASK (  create_problem_default) TYPE (      CreateProblemDefaultAction) NAME (                ) Memory usage 68MB
[DBG][ACT] TASK (       setup_executioner) TYPE (         CreateExecutionerAction) NAME (     Executioner) Memory usage 69MB
[DBG][ACT] TASK (            add_executor) TYPE (               AddExecutorAction) NAME (                ) Memory usage 70MB
[DBG][ACT] TASK (   check_integrity_early) TYPE (            CheckIntegrityAction) NAME (                ) Memory usage 70MB
[DBG][ACT] TASK (  init_displaced_problem) TYPE (    CreateDisplacedProblemAction) NAME (            Mesh) Memory usage 70MB
[DBG][ACT] TASK (        add_aux_variable) TYPE (           DisplayGhostingAction) NAME (            Mesh) Memory usage 70MB
[DBG][ACT] TASK (            add_variable) TYPE (               AddVariableAction) NAME (     Temperature) Memory usage 70MB
[DBG][ACT] TASK (add_external_aux_variables) TYPE (    AddExternalAuxVariableAction) NAME (                ) Memory usage 70MB
[DBG][ACT] TASK (        setup_quadrature) TYPE (           SetupQuadratureAction) NAME (                ) Memory usage 70MB
[DBG][ACT] TASK (         add_user_object) TYPE (           DisplayGhostingAction) NAME (            Mesh) Memory usage 71MB
[DBG][ACT] TASK (         add_user_object) TYPE (         RadiationTransferAction) NAME (vacuum_radiation) Memory usage 71MB
[DBG][ACT] TASK (                  add_ic) TYPE (       AddInitialConditionAction) NAME (        rotor_IC) Memory usage 72MB
[DBG][ACT] TASK (                  add_ic) TYPE (       AddInitialConditionAction) NAME (       vacuum_IC) Memory usage 72MB
[DBG][ACT] TASK (                  add_ic) TYPE (       AddInitialConditionAction) NAME (       stator_IC) Memory usage 72MB
[DBG][ACT] TASK (           ready_to_init) TYPE (                     EmptyAction) NAME (                ) Memory usage 72MB
[DBG][ACT] TASK (           setup_dampers) TYPE (              SetupDampersAction) NAME (                ) Memory usage 72MB
[DBG][ACT] TASK (    setup_residual_debug) TYPE (        SetupResidualDebugAction) NAME (           Debug) Memory usage 72MB
[DBG][ACT] TASK (         copy_nodal_vars) TYPE (             CopyNodalVarsAction) NAME (     Temperature) Memory usage 72MB
[DBG][ACT] TASK (         copy_nodal_vars) TYPE (             CopyNodalVarsAction) NAME (                ) Memory usage 72MB
[DBG][ACT] TASK (     copy_nodal_aux_vars) TYPE (             CopyNodalVarsAction) NAME (                ) Memory usage 72MB
[DBG][ACT] TASK (            add_material) TYPE (               AddMaterialAction) NAME (   thermal_rotor) Memory usage 72MB
[DBG][ACT] TASK (            add_material) TYPE (               AddMaterialAction) NAME (  thermal_vacuum) Memory usage 72MB
[DBG][ACT] TASK (            add_material) TYPE (               AddMaterialAction) NAME (  thermal_stator) Memory usage 73MB
[DBG][ACT] TASK (            add_material) TYPE (               AddMaterialAction) NAME (density_rotor_stator) Memory usage 73MB
[DBG][ACT] TASK (            add_material) TYPE (               AddMaterialAction) NAME (     density_air) Memory usage 73MB
[DBG][ACT] TASK (add_output_aux_variables) TYPE (            MaterialOutputAction) NAME (                ) Memory usage 73MB
[DBG][ACT] TASK (              add_output) TYPE (                SetupDebugAction) NAME (           Debug) Memory usage 73MB
[DBG][ACT] TASK (              add_output) TYPE (              CommonOutputAction) NAME (         Outputs) Memory usage 73MB
[DBG][ACT] TASK (              add_output) TYPE (                 AddOutputAction) NAME (            rays) Memory usage 73MB
[DBG][ACT] TASK (              add_output) TYPE (                 AddOutputAction) NAME (          exodus) Memory usage 73MB
[DBG][ACT] TASK (              add_output) TYPE (                 AddOutputAction) NAME (         console) Memory usage 73MB
[DBG][ACT] TASK (          add_aux_kernel) TYPE (           DisplayGhostingAction) NAME (            Mesh) Memory usage 73MB
[DBG][ACT] TASK (          add_aux_kernel) TYPE (            MaterialOutputAction) NAME (                ) Memory usage 73MB
[DBG][ACT] TASK (                  add_bc) TYPE (         RadiationTransferAction) NAME (vacuum_radiation) Memory usage 73MB
[DBG][ACT] TASK (                  add_bc) TYPE (                     AddBCAction) NAME (temperature_stator) Memory usage 73MB
[DBG][ACT] TASK (              add_kernel) TYPE (                 AddKernelAction) NAME ( heat_conduction) Memory usage 73MB
[DBG][ACT] TASK (              add_kernel) TYPE (                 AddKernelAction) NAME (heat_conduction_time_derivative) Memory usage 73MB
[DBG][ACT] TASK (resolve_optional_materials) TYPE (ResolveOptionalMaterialPropertiesAction) NAME (                ) Memory usage 73MB
[DBG][ACT] TASK (        add_algebraic_rm) TYPE (    CreateDisplacedProblemAction) NAME (            Mesh) Memory usage 73MB
[DBG][ACT] TASK (        add_algebraic_rm) TYPE (          AddRelationshipManager) NAME (                ) Memory usage 73MB
[DBG][ACT] TASK (         add_coupling_rm) TYPE (    CreateDisplacedProblemAction) NAME (            Mesh) Memory usage 73MB
[DBG][ACT] TASK (         add_coupling_rm) TYPE (          AddRelationshipManager) NAME (                ) Memory usage 73MB
[DBG][ACT] TASK (attach_geometric_rm_final) TYPE (          AddRelationshipManager) NAME (                ) Memory usage 73MB
[DBG][ACT] TASK (     attach_algebraic_rm) TYPE (          AddRelationshipManager) NAME (                ) Memory usage 73MB
[DBG][ACT] TASK (      attach_coupling_rm) TYPE (          AddRelationshipManager) NAME (                ) Memory usage 73MB
[DBG][ACT] TASK (  coupling_functor_check) TYPE (      CouplingFunctorCheckAction) NAME (coupling_functor_check) Memory usage 73MB
[DBG][ACT] TASK (delete_remote_elements_after_late_geometric_ghosting) TYPE (         SetupMeshCompleteAction) NAME (            Mesh) Memory usage 73MB
[DBG][ACT] TASK (            init_problem) TYPE (               InitProblemAction) NAME (                ) Memory usage 73MB
[DBG][ACT] TASK (            check_output) TYPE (               CheckOutputAction) NAME (                ) Memory usage 79MB
[DBG][ACT] TASK (         check_integrity) TYPE (            CheckIntegrityAction) NAME (                ) Memory usage 79MB
[DBG][ACT] TASK (               no_action) TYPE (                     EmptyAction) NAME (                ) Memory usage 79MB
[DBG][ACT] TASK (finish_input_file_output) TYPE (                     EmptyAction) NAME (                ) Memory usage 79MB
[DBG][ACT] TASK (add_ray_boundary_condition) TYPE (         RadiationTransferAction) NAME (vacuum_radiation) Memory usage 79MB
[DBG][ACT] Finished executing all actions with memory usage 79MB
```

**Reply by GiudGiud** [reply]
it doesnt cache by default and I dont see it in the parameters of the action. So I think you ll have to modify the code a little bit to get this to work.

add the caching parameters in the code for the action, when it creates the study, right below this line.

https://github.com/idaholab/moose/blob/935a1fe5bcfffb039f8834d02144ce501d5f8db5/modules/heat_conduction/src/actions/RadiationTransferAction.C#L233

**Reply by mrplants** [reply]
Hmm, I'm a little green so a little more assistance for how to proceed would be awesome.  But let me give this a shot.  After modifying the code, I would need to re-`make` moose, right?

**Reply by GiudGiud** [reply]
Yup you need to `make` every time you modify the code.
It should only rebuild a few files at most though (unless you touch a header).

No need to add a parameter to the RadiationTransferAction, just hard-code a `true` to caching segments for now

**Reply by mrplants** [reply]
Oh my gosh I think it's working.  Wow.  Code still running- apparently that caching is a lot.  Here's the line I added:

<img width="348" alt="Screenshot 2023-03-12 at 23 37 49" src="https://user-images.githubusercontent.com/3487464/224602915-2f6a7f59-4d02-46c6-9c92-4bb87c0c312c.png">

**Reply by GiudGiud** [reply]
I'm glad
yeah caching segments (and data on each segment) is a lot

**Comment by mrplants** [comment]
Further question: when I view the exodus output, I don't see any rays reflecting off the surfaces.  I only see them emanating from the patches.  Here is a screenshot:

<img width="874" alt="Screenshot 2023-03-13 at 07 27 32" src="https://user-images.githubusercontent.com/3487464/224690096-5f311853-eaf6-4c41-acd7-408211621cfa.png">

**Reply by mrplants** [reply]
Am I doing something wrong?  Shouldn't these rays be reflecting off the vacuum walls based on emissivity and absorption?

**Reply by mrplants** [reply]
I did a little more digging and I just want to confirm my understanding of the GrayDiffuseRadiation action:

It seems like Moose calculates radiative heat transfer based on material properties and view angles between the surfaces.  The patches emit rays using gaussian quadrature.

However, this model seems like it may break down with highly reflective surfaces like polished aluminum.  If the surface is highly reflective, then shouldn't the patches be modeled based on the angle of incident rays?  In that case, I would expect the radiative heat transfer to be calculated from rays reflected multiple times, rather than having each ray die at its first intersection with a boundary.

Does this make sense?  If so, is there a way to allow RadiationTransferAction to reflect rays a few times before killing them?

**Reply by GiudGiud** [reply]
@loganharbour or @snschune may complement this

I dont think it would break down. The idea here is to get the view factors between the patches on both sides, then perform the linear (nonlinear with T dependent properties) solve to converge the radiation heat exchange. The reflectivity is naturally baked into the amount of energy communicated between surfaces, as a factor on how much energy gets sent back. 

The rays are fixed, they are simply here to compute view factors.

The idea is not to track rays of light dynamically between the two surfaces until they dissipate.

**Reply by mrplants** [reply]
Okay, interesting! Do you have some references that go a little deeper. I still don't fully understand. Specifically, I don't understand how this method accounts for specular reflection. It seems to assume only diffuse reflection.

**Reply by GiudGiud** [reply]
I ll refer to the aforementioned developers for this

**Reply by snschune** [reply]
The net radiation transfer method that is implemented is a fairly standard approach to gray (no frequency dependence), diffuse (all boundaries emit & reflect like Lambert radiators). The case of polished Al would be better described by specular reflection which net radiation transfer typically excludes. The difference is that of a white and specular reflected boundary in radiation transport. Note that most engineering surfaces are some mixture of the two. 

To the question about view factors and rays. @GiudGiud already explained it but I want to add one thing. View factors are essentially first flight surface to surface collision probabilities from the collision probability method in radiation transport. The probability that a photon released in direction Omega makes its first contact with surface patch j. The radiative exchange is can be computed from these first flight exchange probabilities by a little bit of book-keeping and math. That is essentially what @GiudGiud described. 
That bookkeeping includes diffuse reflection, absorption, and radiation. That is why rays are killed at first contact with the next surface...it's because we only want first flight surface-to-surface collision probabilities. 

NOTE: the full problem with specular reflection, energy dependence, and absorption in any media within the cavity can also be solved using the MOOSE based radiation transport application Griffin.
